### PR TITLE
TaskStatusDB add tasks

### LINKED ISSUE #2: TaskStatusDB: Methods to add a task/network of tasks to a database
Make it possible to add `Task`s to the database.

* `add_task(self, task: Task, requirements: Iterable[Task]`: add a single task to database, along with edges to things it depends on. A task with no dependencies should be added with status as `AVAILABLE`, otherwise status should be `BLOCKED`.
* `add_task_network(self, network: nx.DiGraph)`: add an entire graph (with `Task`s as nodes) to the database. Initial status as with `add_task`.

These should probably use some internal methods, rather than having `add_task_network` call `add_task` for each task (which would require a separate transaction with the DB for each task). Adding tasks to the DB should be batched.

### PR DESCRIPTION
**Pull request by dwhswenson**
Resolves #2.

Includes all of #9, so that one should be reviewed first.

NB: Tasks must be added in a reasonable execution order (i.e., topological sort of the DAG), because we use FK constraints on the dependencies table. So no task ID can be added to the dependencies table before it gets added to the tasks table.

### CODE REVIEW

File: exorcist/taskdb.py [hunk 1]
```diff
@@ -1,11 +1,16 @@
...
 # remaining imports are for typing
 from typing import Optional, Iterable
-from .models import TaskStatus
 from os import PathLike
-import networkx as nx
+
+def sqlite_fk_pragma(dbapi_conn, conn_record):
+    cursor = dbapi_conn.cursor()
+    cursor.execute("PRAGMA foreign_keys=ON")
+    cursor.close()
```
Comment (richardjgowers) [inline exorcist/taskdb.py:13]: document what this does and why it's necessary, probably make it hidden behind an underscore

File: exorcist/taskdb.py [hunk 2]
```diff
@@ -188,11 +201,68 @@ def _create_empty_db(metadata, engine):
...
-    def add_task_network(self, task_network: nx.DiGraph):
-        ...
+        Parameters
+        ----------
+        taskid: str
+            the task to add to the database
+        requirements: Iterable[str]
+            taskids that directly block the task to be added (typically,
+            whose outputs are inputs to the task)
+        """
```
Comment (richardjgowers) [inline exorcist/taskdb.py:242]: document all parameters

File: exorcist/taskdb.py [hunk 3]
```diff
@@ -188,11 +201,68 @@ def _create_empty_db(metadata, engine):
...
+    def add_task_network(self, taskid_network: nx.DiGraph, max_tries: int):
+        """Add a network of tasks to the database.
+
+        Parameters
+        ----------
+        taskid_network: nx.Digraph
+            A network with taskids (str) as nodes. Edges in this graph
+            follow the direction of time/flow of information: from earlier
+            tasks to later tasks; from requirements to subsequent.
+        """
```
Comment (richardjgowers) [inline exorcist/taskdb.py:256]: document all parameters

File: exorcist/taskdb.py [hunk 4]
```diff
@@ -28,7 +33,8 @@ class NoStatusChange(Exception):
 
 class AbstractTaskStatusDB(abc.ABC):
     @abc.abstractmethod
-    def add_task(self, taskid: str, requirements: Iterable[str]):
+    def add_task(self, taskid: str, requirements: Iterable[str],
+                 max_tries: int):
```
Comment (richardjgowers) [inline exorcist/taskdb.py:37]: requires docs on max_tries here too

**Review by richardjgowers — CHANGES_REQUESTED** [review]
needs docs on the methods you're adding here.  It's probably worth splitting the sqla class into a separate file too at some point

**Comment by dwhswenson** [comment]
Docs updated. This should be ready for another review.

I don't see a point to splitting the SQLA-based implementation into a separate file. Realistically, it is the only implementation we're likely to do, unless SQLA dies.

**Review by richardjgowers — APPROVED** [review]

**Pull request merged**